File: Beecrowd_Parte2/Eu_Posso_Adivinhar_a_Estrutura_de_Dados_1340.c
/*
Existe uma estrutura de dados do tipo sacola, suportando duas operações:

1 x
Jogue um elemento x na sacola.

2
Tire um elemento da sacola.

Dada uma sequencia de operações que retornam valores, você vai adivinhar a estrutura de dados. É uma pilha (último-dentro, primeiro-fora), uma fila (primeiro-dentro, primeiro-fora), uma fila de prioridade (sempre tire os elementos grandes por primeiro) ou qualquer outra coisa que você dificilmente consegue imaginar!

Entrada
Existem muitos casos de testes. Cada caso de teste começa com a linha contando um único inteiro n (1 <= n <= 1000). Cada uma das seguintes n linhas é um comando do tipo 1, ou um número inteiro 2, seguido de um número inteiro x. Isso significa que depois de executar um comando do tipo 2, obtemos um elemento x sem erros. O valor de x é sempre um número inteiro, positivo e não maior do que 100. O final da entrada é determinado pelo final do arquivo (EOF). O tamanho do arquivo de entrada não excede 1MB.

Saída
Para cada caso de teste, mostre um dos seguintes:

stack
É definitivamente uma pilha.

queue
É definitivamente uma fila.

priority queue
É definitivamente uma fila de prioridade.

impossible
Não pode ser uma pilha, uma fila ou uma fila de prioridade.

not sure
Pode ser mais de uma das três estruturas mencionadas acima.


Exemplo de Entrada	
6
1 1
1 2
1 3
2 1
2 2
2 3
6
1 1
1 2
1 3
2 3
2 2
2 1
2
1 1
2 2
4
1 2
1 1
2 1
2 2
7
1 2
1 5
1 1
1 3
2 5
1 4
2 4

Exemplo de Saída
queue
not sure
impossible
stack
priority queue

*/

#include <stdio.h>
#include <stdbool.h>

#define MAX_N 1005

typedef struct {
    int dados[MAX_N];
    int topo;
} Pilha;

typedef struct {
    int dados[MAX_N];
    int inicio, fim;
} Fila;

typedef struct {
    int dados[MAX_N];
    int tamanho;
} FilaPrioridade;

void inicializarPilha(Pilha *p) {
    p->topo = -1;
}

void inicializarFila(Fila *f) {
    f->inicio = 0;
    f->fim = 0;
}

void inicializarFilaPrioridade(FilaPrioridade *fp) {
    fp->tamanho = 0;
}

void empilhar(Pilha *p, int numero) {
    p->dados[++p->topo] = numero;
}

void desempilhar(Pilha *p) {
    p->topo--;
}

int topo(Pilha *p) {
    return p->dados[p->topo];
}

void enfileirar(Fila *f, int numero) {
    f->dados[f->fim++] = numero;
}

void desenfileirar(Fila *f) {
    f->inicio++;
}

int frente(Fila *f) {
    return f->dados[f->inicio];
}

void enfileirarPrioridade(FilaPrioridade *fp, int numero) {
    fp->dados[fp->tamanho++] = numero;
}

void desenfileirarPrioridade(FilaPrioridade *fp) {
    int maxIndex = 0;
    for (int i = 1; i < fp->tamanho; ++i) {
        if (fp->dados[i] > fp->dados[maxIndex]) {
            maxIndex = i;
        }
    }
    for (int i = maxIndex; i < fp->tamanho - 1; ++i) {
        fp->dados[i] = fp->dados[i + 1];
    }
    fp->tamanho--;
}

int frentePrioridade(FilaPrioridade *fp) {
    int maxIndex = 0;
    for (int i = 1; i < fp->tamanho; ++i) {
        if (fp->dados[i] > fp->dados[maxIndex]) {
            maxIndex = i;
        }
    }
    return fp->dados[maxIndex];
}

int main() {
    bool p, f, fp;
    Pilha pilha;
    Fila fila;
    FilaPrioridade filaPrioridade;
    int n, operacao, numero;

    while (scanf("%d", &n) == 1) {
        inicializarPilha(&pilha);
        inicializarFila(&fila);
        inicializarFilaPrioridade(&filaPrioridade);

        p = true, f = true, fp = true;
        for (int i = 0; i < n; ++i) {
            scanf("%d %d", &operacao, &numero);

            if (operacao == 1) {
                if (p) empilhar(&pilha, numero);
                if (f) enfileirar(&fila, numero);
                if (fp) enfileirarPrioridade(&filaPrioridade, numero);
            } else {
                if (p) {
                    if (pilha.topo == -1 || topo(&pilha) != numero)
                        p = false;
                    else
                        desempilhar(&pilha);
                }

                if (f) {
                    if (fila.inicio == fila.fim || frente(&fila) != numero)
                        f = false;
                    else
                        desenfileirar(&fila);
                }

                if (fp) {
                    if (filaPrioridade.tamanho == 0 || frentePrioridade(&filaPrioridade) != numero)
                        fp = false;
                    else
                        desenfileirarPrioridade(&filaPrioridade);
                }
            }
        }

        if (p && !f && !fp)
            printf("stack\n");
        else if (!p && f && !fp)
            printf("queue\n");
        else if (!p && !f && fp)
            printf("priority queue\n");
        else if (!p && !f && !fp)
            printf("impossible\n");
        else
            printf("not sure\n");
    }

    return 0;
}

File: Beecrowd_Parte2/Eu_Posso_Adivinhar_a_Estrutura_de_Dados_1340.java
/*
Existe uma estrutura de dados do tipo sacola, suportando duas operações:

1 x
Jogue um elemento x na sacola.

2
Tire um elemento da sacola.

Dada uma sequencia de operações que retornam valores, você vai adivinhar a estrutura de dados. É uma pilha (último-dentro, primeiro-fora), uma fila (primeiro-dentro, primeiro-fora), uma fila de prioridade (sempre tire os elementos grandes por primeiro) ou qualquer outra coisa que você dificilmente consegue imaginar!

Entrada
Existem muitos casos de testes. Cada caso de teste começa com a linha contando um único inteiro n (1 <= n <= 1000). Cada uma das seguintes n linhas é um comando do tipo 1, ou um número inteiro 2, seguido de um número inteiro x. Isso significa que depois de executar um comando do tipo 2, obtemos um elemento x sem erros. O valor de x é sempre um número inteiro, positivo e não maior do que 100. O final da entrada é determinado pelo final do arquivo (EOF). O tamanho do arquivo de entrada não excede 1MB.

Saída
Para cada caso de teste, mostre um dos seguintes:

stack
É definitivamente uma pilha.

queue
É definitivamente uma fila.

priority queue
É definitivamente uma fila de prioridade.

impossible
Não pode ser uma pilha, uma fila ou uma fila de prioridade.

not sure
Pode ser mais de uma das três estruturas mencionadas acima.


Exemplo de Entrada	
6
1 1
1 2
1 3
2 1
2 2
2 3
6
1 1
1 2
1 3
2 3
2 2
2 1
2
1 1
2 2
4
1 2
1 1
2 1
2 2
7
1 2
1 5
1 1
1 3
2 5
1 4
2 4

Exemplo de Saída
queue
not sure
impossible
stack
priority queue

*/
import java.util.*;

class Pilha {
    private int[] dados;
    private int topo;

    public Pilha(int tamanhoMaximo) {
        dados = new int[tamanhoMaximo];
        topo = -1;
    }

    public void empilhar(int numero) {
        dados[++topo] = numero;
    }

    public void desempilhar() {
        topo--;
    }

    public int topo() {
        return dados[topo];
    }

    public boolean vazia() {
        return topo == -1;
    }
}

class Fila {
    private int[] dados;
    private int inicio, fim;

    public Fila(int tamanhoMaximo) {
        dados = new int[tamanhoMaximo];
        inicio = 0;
        fim = 0;
    }

    public void enfileirar(int numero) {
        dados[fim++] = numero;
    }

    public void desenfileirar() {
        inicio++;
    }

    public int frente() {
        return dados[inicio];
    }

    public boolean vazia() {
        return inicio == fim;
    }
}

class FilaPrioridade {
    private int[] dados;
    private int tamanho;

    public FilaPrioridade(int tamanhoMaximo) {
        dados = new int[tamanhoMaximo];
        tamanho = 0;
    }

    public void enfileirar(int numero) {
        dados[tamanho++] = numero;
    }

    public void desenfileirar() {
        int maxIndex = 0;
        for (int i = 1; i < tamanho; ++i) {
            if (dados[i] > dados[maxIndex]) {
                maxIndex = i;
            }
        }
        for (int i = maxIndex; i < tamanho - 1; ++i) {
            dados[i] = dados[i + 1];
        }
        tamanho--;
    }

    public int frente() {
        int maxIndex = 0;
        for (int i = 1; i < tamanho; ++i) {
            if (dados[i] > dados[maxIndex]) {
                maxIndex = i;
            }
        }
        return dados[maxIndex];
    }

    public boolean vazia() {
        return tamanho == 0;
    }
}

public class Eu_Posso_Adivinhar_a_Estrutura_de_Dados_1340 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (scanner.hasNextInt()) {
            int n = scanner.nextInt();

            Pilha pilha = new Pilha(n);
            Fila fila = new Fila(n);
            FilaPrioridade filaPrioridade = new FilaPrioridade(n);

            boolean p = true, f = true, fp = true;

            for (int i = 0; i < n; ++i) {
                int operacao = scanner.nextInt();
                int numero = scanner.nextInt();

                if (operacao == 1) {
                    if (p)
                        pilha.empilhar(numero);
                    if (f)
                        fila.enfileirar(numero);
                    if (fp)
                        filaPrioridade.enfileirar(numero);
                } else {
                    if (p) {
                        if (pilha.vazia() || pilha.topo() != numero)
                            p = false;
                        else
                            pilha.desempilhar();
                    }

                    if (f) {
                        if (fila.vazia() || fila.frente() != numero)
                            f = false;
                        else
                            fila.desenfileirar();
                    }

                    if (fp) {
                        if (filaPrioridade.vazia() || filaPrioridade.frente() != numero)
                            fp = false;
                        else
                            filaPrioridade.desenfileirar();
                    }
                }
            }

            if (p && !f && !fp)
                System.out.println("stack");
            else if (!p && f && !fp)
                System.out.println("queue");
            else if (!p && !f && fp)
                System.out.println("priority queue");
            else if (!p && !f && !fp)
                System.out.println("impossible");
            else
                System.out.println("not sure");
        }

        scanner.close();
    }
}
